### DISCUSSION TITLE
Implement a proper import framework

### DISCUSSION THREAD

**Discussion by nielstron** [OpShin/opshin#577] [Ideas / Improvements]
**Is your feature request related to a problem? Please describe.**
Currently only `from x import *` is supported, in a very hacky way, by copying the whole content of that file into the current file.

**Describe the solution you'd like**
OpShin should have a proper, modular import framework.

One way this could be achieved is by copying the imported stuff into a function, "x" which then "returns" all modules that are imported explicitly, i.e. `from x import y, z` --> `def x(): ... return y, z; y, z = x()`

**Alternatives**
Alteranatively, one could do a dependency analysis of required names and fucntions for `y` and `z`, and delete all other members of the module using dead code analysis. It could kind of naturally occur if we still copy the entire module as currently, but change type inference to "forget" everything except for `y` and `z` to be accessed. Then dead code analysis would do the rest.

> [Bug bounty](https://github.com/OpShin/opshin/blob/main/BUG_BOUNTY.md): 2000 ADA

**Comment by chrissiwaffler** [comment]
I think the following problem might also be solved when having a better import framework:

When using 5 files with following structure and splitting the validator logic across the files:
- main.py (import all other files)
- functionality_1.py
- functionality_2.py
- ...

the compiled `.cbor` file increases a lot in size: 47 kB instead of 34 kB all functionality in a single file (_analyzing my personal project_)

when this compiled script is attached to a transaction, it might therefore result in exceeding the maximum transaction size of 16384 B

**Comment by nielstron** [comment]
I think the issue pointed out by @chrissiwaffler should be resolved since #268 , but the main issue is still open.

**Comment by nielstron** [comment]
Another issue is that relative imports don't work as of now. Posting this here for now, but is itself not eligible for the bug bounty.
